### ISSUE TITLE
Lint on hashing smart pointer `as_ptr` with metadata

### ISSUE THREAD

**Issue by edwloef** [rust-lang/rust-clippy#14141]
### What it does

Lint on `Arc::as_ptr(...).hash(state)` and `Rc::as_ptr(...).hash(state)`, and suggest using `Arc::as_ptr(...).addr().hash(state)` and `Rc::as_ptr(...).addr().hash(state)` respectively.

### Advantage

Doing `Arc::as_ptr(&Arc<dyn Trait>).hash(state)` can change behavior run-by-run, as well as between debug and release mode, and is as far as I can tell almost always a bug since hashing metadata doesn't seem often useful to me.

Doing `Arc::as_ptr(...).addr().hash(state)` should almost always be preferrable since `addr` is a noop and this kind of issue is quite tricky to debug. Additionally, `Arc::ptr_eq` compares pointer addresses, not fat pointers.

The same should apply for `Rc` as well, though I haven't personally tried out whether the issue can still occur.

### Drawbacks

It changes behavior, which could be unexpected for people who want to hash the pointer metadata as well. However I believe the changed behavior is what would be commonly expected from the "incorrect" code.

### Example

```rust
pub struct Test(Arc<dyn Debug>);

impl Hash for Test {
    fn hash<H: Hasher>(&self, state: &mut H) {
        Arc::as_ptr(&self.0).hash(state);
    }
}
```

Could be written as:

```rust
pub struct Test(Arc<dyn Debug>);

impl Hash for Test {
    fn hash<H: Hasher>(&self, state: &mut H) {
        Arc::as_ptr(&self.0).addr().hash(state);
    }
}
```

**Comment by samueltardieu** [comment]
I am not sure I understand the issue with `Arc` or `Rc`. Isn't your grudge with the hashing of a fat pointer (the `dyn Debug` in your example) vs. the hashing of its address?

Wouldn't hashing `&0i32 as &dyn Debug` vs. `&0i32` exhibit the same issue?

**Comment by edwloef** [comment]
I guess hashing the fat pointer is the root issue yeah, but I also can't think of a case where that would happen outside of Arc and Rc. If clippy is able to detect all fat pointer hashes that would be cool as well :)
